src/main/java/org/datadog/jenkins/plugins/datadog/transport/HttpSender.java

```diff
@@ -0,0 +1,106 @@
+import java.util.logging.Logger;
+
+public class HttpSender implements Runnable {
+
+    private static final Logger logger = Logger.getLogger(HttpSender.class.getName());
+
+    private final BlockingQueue<HttpMessage> queue;
+    private final HttpErrorHandler errorHandler;
+    private volatile boolean shutdown;
+
+    HttpSender(final int queueSize, final HttpErrorHandler errorHandler) {
+        this(new LinkedBlockingQueue<HttpMessage>(queueSize), errorHandler);
+    }
+
+    HttpSender(final BlockingQueue<HttpMessage> queue, final HttpErrorHandler errorHandler) {
+        this.queue = queue;
+        this.errorHandler = errorHandler;
+    }
+
+    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
+    boolean send(final HttpMessage message){
+        if(!shutdown){
+            queue.offer(message);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void run() {
+        while (!(queue.isEmpty() && shutdown)) {
+            try {
+                final HttpMessage message = queue.poll(1, TimeUnit.SECONDS);
+                if(null != message) {
+                    blockingSend(message);
+                }
+            } catch (final InterruptedException e) {
+                if (shutdown) {
+                    return;
+                }
+            } catch (final Exception e) {
+                errorHandler.handle(e);
+            }
+        }
+    }
+
+    private void blockingSend(HttpMessage message) {
+        HttpURLConnection conn = null;
+        boolean status = true;
+        try {
+            conn = getHttpURLConnection(message.getURL());
+            conn.setRequestMethod(message.getMethod().name());
+            conn.setRequestProperty("Content-Type", message.getContentType());
+            conn.setUseCaches(false);
+            conn.setDoOutput(true);
+
+            final byte[] payload = message.getPayload();
+            final OutputStream outputStream = conn.getOutputStream();
+            outputStream.write(payload);
+            outputStream.close();
+
+            int httpStatus = conn.getResponseCode();
+            if(httpStatus >= 400) {
+                logger.severe("Failed to send HTTP request: "+message.getMethod()+" "+ message.getURL()+ " - Status: HTTP "+conn.getResponseCode());
+                status = false;
+            }
+        } catch (Exception ex) {
+            errorHandler.handle(ex);
+            try {
+                if(conn != null) {
+                    logger.severe("Failed to send HTTP request: "+message.getMethod()+" "+ message.getURL()+ " - Status: HTTP "+conn.getResponseCode());
+                }
+            } catch (IOException ioex) {
+                errorHandler.handle(ioex);
+            }
+            status = false;
+        } finally {
+            if(conn != null) {
+                conn.disconnect();
+            }
+        }
+    }
+
+
+
+    boolean isShutdown() {
+        return shutdown;
```

Comment:
With the current code, `isShutdown` could return `false` while the sender is still actively sending messages. Not sure if that is a concern.

Reply:
I guess you mean the `client` is still actively sending messages. I think it should be fine for the moment because the shutdown only could happen when Jenkins is restarted, and all running pipelines must finish before the restart.